### ISSUE TITLE
__cxa_allocate_exception does not provide required alignment on exception object

### ISSUE THREAD

**Issue by emaste** [pathscale/libcxxrt#38]
LLVM libunwind is about to require `struct _Unwind_Exception` to have ABI-maximum alignment and GCC's already does. The memory currently returned by __cxa_allocate_exception may be under-aligned.

**Comment by emaste** [comment]
Patch for FreeBSD at https://reviews.freebsd.org/D7263

**Comment by emaste** [comment]
Oops, it's not so simple; the misalignment comes cxxabi.h; the `_Unwind_Exception unwindHeader` in `struct __cxa_exception` is at offset 0x58, and `sizeof(__cxa_exception)` is 0x78.

**Comment by emaste** [comment]
On FreeBSD amd64 `struct __cxa_exception` looks like this:
```
0x00 uintptr_t referenceCount
0x08 std::type_info *exceptionType
0x10 void (*exceptionDestructor) (void *)
0x18 unexpected_handler unexpectedHandler
0x20 terminate_handler terminateHandler
0x28 __cxa_exception *nextException
0x30 int handlerCount
0x34 int handlerSwitchValue
0x38 const char *actionRecord
0x40 const char *languageSpecificData
0x48 void *catchTemp
0x50 void *adjustedPtr
0x58 _Unwind_Exception unwindHeader
0x78
```

**Comment by emaste** [comment]
However, see also https://llvm.org/bugs/show_bug.cgi?id=24604 -- something like the patch in review D7263 above is probably needed for the 32-bit case.

**Comment by jsonn** [comment]
I don't think this dance is necessary. If the system malloc doesn't return memory aligned enough, you have far bigger problems.

**Comment by emaste** [comment]
@jsonn note there are two separate issues here.

On amd64/x86-64 calloc returns an adequately aligned pointer (16-byte alignment), but the `_Unwind_Exception` is misaligned (it requires 16-byte alignment, but its offset within `__exa_exception` is not on a 16-byte boundary).

The linked LLVM bug report is for 32-buit builds, and claims __cxa_exception requires 16-byte alignment. On FreeBSD/i386 AFAICT the maximal alignment is indeed 16 bytes, and malloc returns 16-byte aligned addresses for size >= 16, so this 2nd issue is not a concern for FreeBSD.

**Comment by emaste** [comment]
Workaround committed to FreeBSD: https://reviews.freebsd.org/rS303157

**Comment by emaste** [comment]
Also note that libstdc++ (libsupc++) had the same bug originally, reported here: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=38732

The libstdc++ solution was to introduce
```
struct __cxa_refcounted_exception
{
    ... referenceCount;
    __cxa_exception exc;
}
```
leaving __cxa_exception as it was prior to the introduction of referenceCount, and we probably need to do the same. Because __cxa_exception has a 16-byte alignment requirement the compiler adds the necessary padding.
